src/Ens.ts

```diff
@@ -0,0 +1,594 @@
+    const returnee = (await this.records(domain, [key]))[key];
+    if (!returnee) {
+      throw new ResolutionError(ResolutionErrorCode.RecordNotFound, {
+        domain,
+        recordName: key,
+      });
+    }
+    return returnee;
+  }
+
+  async records(
+    domain: string,
+    keys: string[],
+  ): Promise<Record<string, string>> {
+    const values = await Promise.all(
+      keys.map(async (key) => {
+        if (key.startsWith('crypto.')) {
+          const ticker = key.split('.')[1];
+          return await this.addr(domain, ticker);
+        }
+        if (key === 'ipfs.html.value' || key === 'dweb.ipfs.hash') {
+          return await this.getContentHash(domain);
+        }
+        const ensRecordName = this.fromUDRecordNameToENS(key);
+        return await this.getTextRecord(domain, ensRecordName);
+      }),
+    );
+    return constructRecords(keys, values);
+  }
+
+  async reverse(
+    address: string,
+    currencyTicker: string,
+  ): Promise<string | null> {
+    if (currencyTicker != 'ETH') {
+      throw new Error(`Ens doesn't support any currency other than ETH`);
+    }
+
+    if (address.startsWith('0x')) {
+      address = address.substr(2);
+    }
+
+    const reverseAddress = address + '.addr.reverse';
+    const nodeHash = this.namehash(reverseAddress);
+    const resolverAddress = await this.resolver(reverseAddress).catch(
+      (err: ResolutionError) => {
+        if (err.code === ResolutionErrorCode.UnspecifiedResolver) {
+          return null;
+        }
+        throw err;
+      },
+    );
+
+    if (isNullAddress(resolverAddress)) {
+      return null;
+    }
+
+    const resolverContract = new EthereumContract(
+      resolverInterface(resolverAddress, EthCoinIndex),
+      resolverAddress,
+      this.provider,
+      this.proxyServiceApiKey,
+    );
+
+    return await this.resolverCallToName(resolverContract, nodeHash);
+  }
+
+  async reverseOf(
+    address: string,
+    location?: UnsLocation,
+  ): Promise<string | null> {
+    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
+      methodName: 'reverseOf',
+    });
+  }
+
+  async getTokenUri(domain: string): Promise<string> {
+    const tokenId = this.namehash(domain);
+    const isWrappedDomain = await this.determineIsWrappedDomain(tokenId);
+    if (isWrappedDomain) {
+      return `https://metadata.ens.domains/${this.networkName}/${this.nameWrapperContract.address}/${tokenId}`;
+    }
+
+    const hashedLabel = labelNameHash(domain);
+    return `https://metadata.ens.domains/${this.networkName}/${this.baseRegistrarContract.address}/${hashedLabel}`;
+  }
+
+  async isAvailable(domain: string): Promise<boolean> {
+    return !(await this.isRegistered(domain));
+  }
+
+  async registryAddress(domain: string): Promise<string> {
+    return this.registryContract.address;
+  }
+
+  async isRegistered(domain: string): Promise<boolean> {
+    const address = await this.owner(domain);
```

Comment:
maybe this is fine but should we call `recordExists` in registry contract instead?

Reply:
Does recordExists mean the same if it's registered?

Docs say:
> Returns true if node exists in this ENS registry. This will return false for records that are in the legacy ENS registry but have not yet been migrated to the new one.

Could it exist in the registry and not be registered? or does it only mean that when it's registered then it exists?